ulib/FStar.List.Tot.Properties.fst

```diff
@@ -596,13 +596,23 @@ let rec partition_count_forall #a f l= match l with
 
 let rec mem_subset (#a: eqtype) (la lb: list a)
     : Lemma (requires (forall x. mem x la ==> mem x lb))
-            (ensures (subset la lb)) =
+            (ensures (subset la lb))
+            [SMTPat (subset la lb)] =
   match la with
   | [] -> ()
   | hd :: tl -> mem_subset tl lb
 
+let rec subset_mem (#a: eqtype) (la lb: list a)
+    : Lemma (requires (subset la lb))
+            (ensures (forall x. mem x la ==> mem x lb))
+            [SMTPat (subset la lb)] =
```

Comment:
Hi @chandradeepdey, this pattern is not optimal for the lemma, since it is not goal-directed. Usually, we pick patterns that mention terms in the goals (i.e., `ensures`) so that Z3 instantiates the lemma when it needs to prove the postcondition.

Rest of the changes look good to me. So you absolutely need this smtpat? E.g., could you remove the pattern, and invoke the lemma manually?

Reply:
(If you prefer, I can also do this change in your branch, let me know.)

Reply:
Oh, okay. What I wanted was `subset la lb <==> (forall x. mem x la ==> mem x lb)` to be put in the context whenever `subset la lb` is mentioned.

I was using `subset_mem` to send a subset as a function argument. I don't need this new one, but thought it would be useful to have the reverse, i.e. get the information out of a function parameter.

Sure, you can make changes.

Reply:
Thanks again!